searchcore/src/vespa/searchcore/proton/server/documentbucketmover.h

```diff
@@ -6,48 +6,113 @@
+  */
+class BucketMover
 {
+public:
+    using MoveOperationUP = std::unique_ptr<MoveOperation>;
+    using IDestructorCallback = vespalib::IDestructorCallback;
+    using IDestructorCallbackSP = std::shared_ptr<IDestructorCallback>;
+    struct MoveKey
+    {
+        using Timestamp = storage::spi::Timestamp;
+        MoveKey(uint32_t lid, const document::GlobalId &gid, Timestamp timestamp) noexcept
+            : _lid(lid),
+              _gid(gid),
+              _timestamp(timestamp)
+        { }
+
+        uint32_t           _lid;
+        document::GlobalId _gid;
+        Timestamp          _timestamp;
+    };
+
+    BucketMover(const document::BucketId &bucket, const MaintenanceDocumentSubDB *source, uint32_t targetSubDbId,
+                IDocumentMoveHandler &handler, BucketDBOwner &bucketDb) noexcept;
+    BucketMover(BucketMover &&) noexcept = default;
+    BucketMover & operator=(BucketMover &&) noexcept = delete;
+    BucketMover(const BucketMover &) = delete;
+    BucketMover & operator=(const BucketMover &) = delete;
+
+    // TODO remove once we have swaitched bucket move job
+    bool moveDocuments(size_t maxDocsToMove, IMoveOperationLimiter &limiter);
+
+    /// Must be called in master thread
+    std::pair<std::vector<MoveKey>, bool> getKeysToMove(size_t maxDocsToMove) const;
+    /// Call from any thread
+    std::vector<MoveOperationUP> createMoveOperations(const std::vector<MoveKey> & toMove);
+    /// Must be called in master thread
+    void moveDocuments(std::vector<MoveOperationUP> moveOps, IDestructorCallbackSP onDone);
+
+    const document::BucketId &getBucket() const { return _bucket; }
+    void cancel() { setBucketDone(); }
+    void setBucketDone() { _bucketDone = true; }
+    bool bucketDone() const { return _bucketDone; }
+    const MaintenanceDocumentSubDB * getSource() const { return _source; }
+    /// Must be called in master thread
+    void updateLastValidGid(const document::GlobalId &gid) {
+        _lastGid = gid;
+        _lastGidValid = true;
+    }
 private:
-    IMoveOperationLimiter          &_limiter;
-    document::BucketId              _bucket;
     const MaintenanceDocumentSubDB *_source;
-    uint32_t                        _targetSubDbId;
     IDocumentMoveHandler           *_handler;
     BucketDBOwner                  *_bucketDb;
+    const document::BucketId        _bucket;
+    const uint32_t                  _targetSubDbId;
+
     bool                            _bucketDone;
     document::GlobalId              _lastGid;
     bool                            _lastGidValid;
 
-    bool moveDocument(uint32_t lid,
-                      const document::GlobalId &gid,
-                      storage::spi::Timestamp timestamp);
+    void moveDocument(MoveOperationUP moveOp, IDestructorCallbackSP onDone);
+    MoveOperationUP createMoveOperation(const MoveKey & key);
+};
+}
 
-    void setBucketDone();
+/**
+ * Class used to move all documents in a bucket from a source sub database
```

Comment:
Ensure that this description and the description for BucketMover are up to date.